internal/handler/msg_createuser.go

```diff
@@ -108,65 +103,39 @@ func (h *Handler) MsgCreateUser(ctx context.Context, msg *wire.OpMsg) (*wire.OpM
 
 	common.Ignored(document, h.L, "writeConcern", "authenticationRestrictions", "comment")
 
-	defMechanisms := must.NotFail(types.NewArray("SCRAM-SHA-1", "SCRAM-SHA-256"))
+	allMechanisms := must.NotFail(types.NewArray("SCRAM-SHA-1", "SCRAM-SHA-256"))
 
-	mechanisms, err := common.GetOptionalParam(document, "mechanisms", defMechanisms)
+	mechanisms, err := common.GetOptionalParam(document, "mechanisms", allMechanisms)
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
-	var credentials *types.Document
+	if mechanisms.Len() == 0 {
+		return nil, handlererrors.NewCommandErrorMsg(
+			handlererrors.ErrBadValue,
+			"mechanisms field cannot be empty",
+		)
+	}
 
+	var password string
 	if document.Has("pwd") {
-		pwdi := must.NotFail(document.Get("pwd"))
-		pwd, ok := pwdi.(string)
+		pwd := must.NotFail(document.Get("pwd"))
+		_, ok := pwd.(string)
 
 		if !ok {
 			return nil, handlererrors.NewCommandErrorMsg(
 				handlererrors.ErrTypeMismatch,
 				fmt.Sprintf("BSON field 'createUser.pwd' is the wrong type '%s', expected type 'string'",
-					handlerparams.AliasFromType(pwdi),
+					handlerparams.AliasFromType(pwd),
 				),
 			)
 		}
 
-		credentials, err = makeCredentials(mechanisms, username, pwd)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	id := uuid.New()
-	saved := must.NotFail(types.NewDocument(
-		"_id", dbName+"."+username,
-		"credentials", credentials,
-		"user", username,
-		"db", dbName,
-		"roles", types.MakeArray(0),
-		"userId", types.Binary{Subtype: types.BinaryUUID, B: must.NotFail(id.MarshalBinary())},
-	))
-
-	adminDB, err := h.b.Database("admin")
-	if err != nil {
-		return nil, lazyerrors.Error(err)
-	}
-
-	users, err := adminDB.Collection("system.users")
-	if err != nil {
-		return nil, lazyerrors.Error(err)
+		password = pwd.(string)
 	}
 
-	_, err = users.InsertAll(ctx, &backends.InsertAllParams{
-		Docs: []*types.Document{saved},
-	})
+	err = backends.CreateUser(ctx, h.b, mechanisms, dbName, username, password)
 	if err != nil {
-		if backends.ErrorCodeIs(err, backends.ErrorCodeInsertDuplicateID) {
```

Comment:
:warning: this error must be returned to keep compatibility.

The issue https://github.com/FerretDB/FerretDB/issues/4243 states that the only case when you don't want to return error is when `--setup-username` flag is passed. Which I don't spot in this codebase.

You should run integration tests to ensure your changes don't introduce regression :hugs:
For example, you can run tests for just `task test-integration-postgresql TEST_RUN=User`. Any failing test is an indication that the change introduce regression.

Reply:
All are passing except for https://github.com/FerretDB/FerretDB/actions/runs/8783663803/job/24100295498?pr=4248#step:10:91. See comment [here](https://github.com/FerretDB/FerretDB/pull/4248#issuecomment-2069245024) as well

Reply:
> All are passing except for https://github.com/FerretDB/FerretDB/actions/runs/8783663803/job/24100295498?pr=4248#step:10:91.

That's no longer the case, there are a lot more tests failing in the most up to date CI run. See several failed tests in CreateUser and some in UpdateUser. I suggest running tests more frequently.